include/mixer.h

```diff
@@ -106,9 +109,11 @@ typedef SpeexResamplerState_ SpeexResamplerState;
 
 class MixerChannel {
 public:
-	MixerChannel(MIXER_Handler _handler, const char *name);
+	MixerChannel(MIXER_Handler _handler, const char *name,
+	             const std::set<ChannelFeature> features);
```

Comment:
when taking in read-only compound and class types, always use a reference: (`const std::set<ChannelFeature> & features`)

Even if a temporary or literal is passed in (which might not be held in a variable), the compiler still needs to construct that object anyway before passing it in (as a copy or a reference). The benefit is that, the const-ref scenario works for literals or variables held outside the function, so it's always the most efficient. 
(I believe PVS Studio will flags these too).

Reply:
Yeah, it definitely makes sense for containers and complex structs if that's the case.

Not sure about C++, but I'd think C compilers would optimise away most pass-struct-by-value cases, no? I know that's what Nim does under the hood (it transpiles to C), and some people voiced their concerns about it, but the language's creator basically claimed that the compiler is clever enough to optimise the copying away, or even inline the functions where it makes a real difference in performance. But that's subtly different, because I'm sure he has tweaked the Nim->C transpiler to generate optimisable code in the first place.

But don't get me wrong, I don't want to change the general coding style in the project, I'll use references consistently. I'm just wondering if the C++ compiler can optimise these cases away (maybe not, as C++ is hugely more complex than C).

Reply:
For reference, I was referring to this:

https://en.wikipedia.org/wiki/Copy_elision

https://bastian.rieck.me/blog/posts/2015/return_value_optimization/

https://en.cppreference.com/w/cpp/language/copy_elision

The rules seem to be quite complicated and compiler specific to some extent (I've just skimmed these articles, there's tons of subtle detail there).

Reply:
Yup, if the compiler can eliminate temporaries, it will.

A good example to show it is return values:
```
std::vector<int> get_things() {
   std::vector things = {}; // temporary
   ... Populate things
   return things; // return by value
}

auto t = get_things();
```

The compiler will use `std::move()` under the hood to make sure `t` comes to life with the content from `things`, without any copies being created or having to recopy values out of the temporary vector into the assigned vector. 

For the `set` in here:
```
MixerChannel(MIXER_Handler _handler, const char *name,
	             const std::set<ChannelFeature> features);
```

That same scenario temporary-elimintation happens here when I pass a literal set (which is a temporary) in, like: `MixerChannel(..., ..., {1, 2, 4})`

But what happens if I'm passing in a longer lived (non-temporary) variable?

```
auto options = {1, 2, 4};
MixerChannel(..., ..., options);
// More code here, which maybe also uses or changes options
```

In this case, a new set is created inside the function (features) and assigned the values from options. So two copies exist.  

When we change it to take in a const &, the first literal temporary scenario still happens (and features is just an alias to that literal temporary), but in this second case, features also gets to be an alias to options. So it's the best of both worlds, without drawbacks. 

Of course.. these sets are tiny, but it's just a handy pattern to use all the time. 

Some good comments here and a bit more details: https://stackoverflow.com/questions/14856344/when-should-stdmove-be-used-on-a-function-return-value

Reply:
Poke around in godbolt with some comparison code (and -O2 optimizations), and the asm results should reveal what you're seeking!

Reply:
Thanks for the explanation; I had a vague idea that something like this might be happening, but your explanations illustrated the concepts clearly.